pkg/provider/chef/chef.go

```diff
@@ -0,0 +1,199 @@
+)
+
+const (
+	errChefStore                             = "received invalid Chef SecretStore resource: %w"
+	errUnexpectedStoreSpec                   = "unexpected store spec"
+	errMissingName                           = "missing Name"
+	errMissingBaseURL                        = "missing BaseURL"
+	errMissingAuth                           = "cannot initialize Chef Client: no valid authType was specified"
+	errMissingPublicKey                      = "missing Public Key"
+	errInvalidClusterStoreMissingPKNamespace = "invalid ClusterSecretStore: missing Chef SecretKey Namespace"
+	errFetchK8sSecret                        = "could not fetch SecretKey Secret: %w"
+	errChefInvalidURL                        = "unable to parse URL: %w"
+	errChefInvalidName                       = "invalid name: allowed values are lowecase letters, numbers, hyphens and underscores"
+	errChefClient                            = "unable to create chef client: %w"
+)
+
+type Providerchef struct {
+	chefClient *chef.Client
+}
+
+// https://github.com/external-secrets/external-secrets/issues/644
+var _ v1beta1.SecretsClient = &Providerchef{}
+var _ v1beta1.Provider = &Providerchef{}
+
+func init() {
+	v1beta1.Register(&Providerchef{}, &v1beta1.SecretStoreProvider{
+		Chef: &v1beta1.ChefProvider{},
+	})
+}
+
+func (providerchef *Providerchef) NewClient(ctx context.Context, store v1beta1.GenericStore, kube kclient.Client, namespace string) (v1beta1.SecretsClient, error) {
+	storeSpec := store.GetSpec()
+	if storeSpec == nil || storeSpec.Provider.Chef == nil {
+		return nil, errors.New(errUnexpectedStoreSpec)
+	}
+	chefSpec := storeSpec.Provider.Chef
+	if chefSpec.Name == "" {
+		return nil, fmt.Errorf(errMissingName)
+	}
+	if chefSpec.BaseURL == "" {
+		return nil, fmt.Errorf(errMissingBaseURL)
+	}
+	if chefSpec.Auth == nil {
+		return nil, fmt.Errorf(errMissingAuth)
+	}
+	if chefSpec.Auth.SecretRef.SecretKey.Key == "" {
+		return nil, fmt.Errorf(errMissingPublicKey)
+	}
+	credentialsSecret := &corev1.Secret{}
+	objectKey := types.NamespacedName{
+		Name:      chefSpec.Auth.SecretRef.SecretKey.Name,
+		Namespace: namespace,
+	}
+	if store.GetObjectKind().GroupVersionKind().Kind == v1beta1.ClusterSecretStoreKind {
+		objectKey.Namespace = *chefSpec.Auth.SecretRef.SecretKey.Namespace
+	} else {
+		return nil, fmt.Errorf(errInvalidClusterStoreMissingPKNamespace)
+	}
+
+	err := kube.Get(ctx, objectKey, credentialsSecret)
+	if err != nil {
+		return nil, fmt.Errorf(errFetchK8sSecret, err)
+	}
+
+	publickey := credentialsSecret.Data[chefSpec.Auth.SecretRef.SecretKey.Key]
+	if (publickey == nil) || (len(publickey) == 0) {
+		return nil, fmt.Errorf(errMissingPublicKey)
+	}
+
+	client, cerr := chef.NewClient(&chef.Config{
+		Name:    chefSpec.Name,
+		Key:     string(publickey),
+		BaseURL: chefSpec.BaseURL,
+	})
+	if cerr != nil {
+		return nil, fmt.Errorf(errChefClient, err)
+	}
+	providerchef.chefClient = client
+	return providerchef, nil
+}
+
+// TO be implemented
```

Comment:
what does this comment relate to?

Reply:
the above comment line indicates to properly implement all the methods of `type SecretsClient interface` 
Currently we are just returning required fields, more to be done on next PR.